File: js/src/ingest/embedder.ts
// The model seam and the write-path vector guard. Deliberately SEPARATE from
// ingest.ts, which loads the native FFI core at import time: the contract a
// provider has to satisfy, and the check that keeps a degenerate vector out of
// the corpus, are pure TypeScript and stay importable and unit-tested with no
// native library built (ADR-0014 R2). Go's twin is golang/ingest/embedder.go.

/** A dense embedder: text in, one vector out. Injected by the caller.
 *
 * ASYNC-CAPABLE (ADR-0014 R2). The old `=> number[]` was unsatisfiable by any
 * network- or model-backed provider: `tsc --strict` rejected every async
 * embedder outright, and casting past it put a `Promise` in `row.vector` (which
 * serialises to `{}` for the FFI store) while the rejection went unhandled, so a
 * failed model call was invisible to `ingest()`. A synchronous embedder — the
 * hermetic fakes — is still valid; `ingest` awaits either.
 *
 * Failure is expressed by REJECTING (or throwing), never by a placeholder
 * vector: a zero vector is not an error value, it is a valid embedding of
 * something, and once written it is indistinguishable from a document that
 * genuinely embeds near the origin. */
export type Embedder = (text: string) => number[] | Promise<number[]>;

/** Is `vec` the all-zeros vector — an embedding carrying no signal at all?
 *  Twin of the Python reference `citenexus.testing.fakes.is_zero_vector`,
 *  promoted from test-only helper to write-path guard. */
export function isZeroVector(vec: readonly number[]): boolean {
  return vec.every((v) => v === 0);
}

/** Reject a vector that must never become an evidence-unit row — the belt to the
 *  async seam's braces. Even a provider that reports success must not be able to
 *  poison the corpus with a degenerate vector.
 *
 *  `dim` is the dimensionality established by this ingest run (0 for the first
 *  vector, which defines it). */
export function checkVector(euId: string, vec: unknown, dim: number): number[] {
  if (!Array.isArray(vec) || vec.some((v) => typeof v !== "number")) {
    throw new TypeError(`ingest: embedder returned a non-vector for ${euId}`);
  }
  const v = vec as number[];
  if (v.length === 0) {
    throw new Error(`ingest: embedder returned an empty vector for ${euId}`);
  }
  if (dim > 0 && v.length !== dim) {
    throw new Error(
      `ingest: embedder returned a ${v.length}-dim vector for ${euId}; this run is ${dim}-dim`,
    );
  }
  if (v.some((x) => !Number.isFinite(x))) {
    throw new Error(`ingest: embedder returned a non-finite vector for ${euId}`);
  }
  if (isZeroVector(v)) {
    throw new Error(
      `ingest: embedder returned the zero vector for ${euId} — it carries no signal ` +
        `and would rank meaninglessly against every query`,
    );
  }
  return v;
}

File: golang/ingest/embedder.go
// The model seam and the write-path vector guard. Deliberately NOT behind the
// `citenexus_ffi` build tag: the contract a provider has to satisfy, and the
// check that keeps a degenerate vector out of the corpus, are pure Go. Keeping
// them here means both are compiled and unit-tested in the default build, with
// no native library and no C toolchain (ADR-0014 R2).

package ingest

import "fmt"

// Embedder turns chunk text into a dense vector, or reports why it could not.
// It is injected so the core owns orchestration and CiteNexus bundles no models.
//
// The error return is the contract, not decoration (ADR-0014 R2): a model that
// times out, refuses, rate-limits or runs out of memory must be able to SAY so.
// Without it the only things an implementation can return are a zero vector, a
// nil slice, or a panic — and a zero vector is not an error value. It is a valid
// embedding of something, so once written it is indistinguishable from a
// document that genuinely embeds near the origin, and retrieval scores it 0.0000
// with no error and no flag. That is a wrong the library cannot see.
//
// An implementation MUST return a non-nil error rather than a placeholder vector
// whenever it did not actually embed the text.
type Embedder interface {
	Embed(text string) ([]float64, error)
}

// EmbedderFunc adapts a plain function to Embedder (the http.HandlerFunc
// pattern), so a client that already has the right shape plugs straight in:
//
//	client := models.NewOpenAIEmbedding(baseURL, model, transport)
//	rows, err := ingest.Ingest(store, data, "pdf", "doc-1", ingest.EmbedderFunc(client.EmbedQuery))
type EmbedderFunc func(text string) ([]float64, error)

// Embed calls f.
func (f EmbedderFunc) Embed(text string) ([]float64, error) { return f(text) }

// isZeroVector reports whether vec carries no signal at all. Twin of the Python
// reference citenexus.testing.fakes.is_zero_vector, promoted here from test-only
// helper to write-path guard.
func isZeroVector(vec []float64) bool {
	for _, v := range vec {
		if v != 0 {
			return false
		}
	}
	return true
}

// checkVector rejects a vector that must never become an Evidence-Unit row. It
// is the belt to the Embedder error return's braces: even with a seam that CAN
// report failure, a provider that returns a degenerate vector without erroring
// must not poison the corpus.
//
// dim is the dimensionality already established by this ingest run (0 for the
// first vector, which defines it). Three rejections:
//
//   - empty/nil — no vector at all; unguarded it becomes a dimension mismatch
//     deep inside the store, misattributed to storage rather than to the model.
//   - wrong dimension — the run's vectors must be mutually comparable.
//   - all zeros — the silent poison ADR-0014 names. Cosine against it does not
//     raise, it just ranks meaninglessly.
func checkVector(euID string, vec []float64, dim int) error {
	if len(vec) == 0 {
		return fmt.Errorf("ingest: embedder returned an empty vector for %s", euID)
	}
	if dim > 0 && len(vec) != dim {
		return fmt.Errorf(
			"ingest: embedder returned a %d-dim vector for %s; this run is %d-dim",
			len(vec), euID, dim,
		)
	}
	if isZeroVector(vec) {
		return fmt.Errorf(
			"ingest: embedder returned the zero vector for %s — it carries no signal and "+
				"would rank meaninglessly against every query", euID,
		)
	}
	return nil
}
